#414: Added examples and explanations for features in documentation (#415)
exasol/notebook-connector: doc/user_guide/slc-examples.rst

Script Language Container (SLC) Examples

> ##########################################

Reviewer: ```suggestion
Script Language Container (SLC) Examples
########################################
```

Script Language Containers (SLCs) are Linux container based Python
environments that
Exasol UDF scripts execute inside.  By building a custom SLC you can include
any Python package your UDFs depend on without modifying the database server.
The Notebook Connector automates the three-step process: register → build &
upload → activate.

**Required extra:** ``pip install "notebook-connector[slc]"``

Step 1 – Register an SLC instance

> ***********************************

Reviewer: ```suggestion
Step 1 – Register an SLC Instance
*********************************
```

Call ``ScriptLanguageContainer.create()`` once per SLC.  This method:

* Clones the `script-languages-release <https://github.com/exasol/script-languages-release>`_
  git repository to the local file system.
* Saves the chosen ``flavor`` in the SCS under the given ``name``.
* Derives a unique language alias (``CUSTOM_SLC_<NAME>``) that will be used
  when activating the container in the database.

The ``name`` must be unique across all SLCs registered in the same SCS, and
the ``flavor`` must match a directory name inside the cloned repository (e.g.
``python3-ds-EXASOL-7.1.0``).

.. code-block:: python

    from exasol.nb_connector.slc import ScriptLanguageContainer

    ScriptLanguageContainer.create(
        secrets=my_secrets,
        name="my_slc",
        flavor="python3-ds-EXASOL-7.1.0",
    )
    # The resulting language alias will be: CUSTOM_SLC_MY_SLC

Step 2 – Build and upload the SLC

> ***********************************

Reviewer: ```suggestion
Step 2 – Build and Upload the SLC
*********************************
```

Instantiate ``ScriptLanguageContainer`` with the same ``name`` used during
``create()``.  Calling ``deploy()`` builds the container image from the local
clone, packages it as a tar archive, streams the archive to BucketFS, and
stores the resulting language activation definition in the SCS.

.. code-block:: python

    from exasol.nb_connector.slc import ScriptLanguageContainer

    slc = ScriptLanguageContainer(secrets=my_secrets, name="my_slc")
    slc.deploy()

Step 3 – Activate the language in the database

> ************************************************

Reviewer: ```suggestion
Step 3 – Activate the Language in the Database
**********************************************
```

Script Language Container (SLC) Examples
##########################################

> Script Language Containers (SLCs) are Docker-based Python environments that

Reviewer 2: ```suggestion
Script Language Containers (SLCs) are Linux container based Python environments that
```

Author: Applied

the ``flavor`` must match a directory name inside the cloned repository (e.g.
``python3-ds-EXASOL-7.1.0``).

.. code-block:: python

    from exasol.nb_connector.slc import ScriptLanguageContainer

    ScriptLanguageContainer.create(
        secrets=my_secrets,
        name="my_slc",
        flavor="python3-ds-EXASOL-7.1.0",
    )
    # The resulting language alias will be: CUSTOM_SLC_MY_SLC

Step 2 – Build and upload the SLC
***********************************

Instantiate ``ScriptLanguageContainer`` with the same ``name`` used during
``create()``.  Calling ``deploy()`` builds the container image from the local
clone, packages it as a tar archive, and streams the archive to BucketFS.
It then registers the language alias in the database so the container is
available for UDF execution.

Pass an open database connection and an open BucketFS bucket; the helper
functions from :doc:`database-connection-examples` and
:doc:`bucketfs-examples` create them from the SCS automatically.

.. code-block:: python

    from exasol.nb_connector.connections import open_pyexasol_connection, open_bucketfs_bucket

    slc = ScriptLanguageContainer(secrets=my_secrets, name="my_slc")
    slc.deploy(

>         connection=open_pyexasol_connection(my_secrets),

Reviewer 2: Is this really needed, I think deploy doesn't need any parameters

Author: Simplified the example to plain `slc.deploy()` with no unnecessary parameters.
